File: packages/perk-dev/src/perk_dev/prose_review/assembly.py
"""Assembly-wide preview composition over one stable catalog/source generation.

:class:`AssemblyRenderer` resolves one validated assembly/scenario pair from an explicit
:class:`~perk_dev.prose_review.catalog.CatalogSnapshot`, applies scenario-default presentation
with nullable caller overrides, validates path-keyed workspace buffers, and returns every
authored layer exactly once in authored order. There is no session-shape parameter and no
layer filtering: a shape is navigation/provenance only, and an ``optional`` layer stays
present with an explicit presence marker. Presentation toggles change only the top-level
:class:`ResolvedPresentation` echo — the per-layer tuple is byte-identical across toggle
values, so clients derive display visibility from ``visibility_control`` plus the resolved
booleans and contradictory wire states are impossible.

Content composition is family-exact and never executes repository code:

- Prompt-root Markdown (``prompt_template_name``) is gated by the frozen-grammar preview scan
  (subset membership, no includes, mapping-only identifiers) before the unchanged production
  ``perk.prompts.render_text`` seam ever compiles it — editable text can never reach Jinja
  with calls/filters/loops/includes or engine-global identifiers, and can never trigger the
  packaged ``prompts_dir()`` include loader.
- Other Markdown is exact raw source, never parsed as Jinja.
- Code-owned layers extract every catalog fragment in order through the existing
  SourceAdapter selector authority (one parse / one helper invocation per layer).
- External boundaries become typed owner/kind placeholders.

Expected source/gate/template/selector failures become one typed
:class:`FailedAssemblyLayer` for only the responsible authored position while every sibling
remains in the result; unexpected invariant failures stay loud. The HTTP caller owns
stabilizing the snapshot and canonical source bytes against saves (the source transaction in
``web.py``); this module stays pure over its inputs plus contained canonical reads.
"""

from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from jinja2 import TemplateError

from perk.prompts import render_text
from perk_dev.prompt_grammar import scan_template
from perk_dev.prose_map.catalog import prompt_template_name
from perk_dev.prose_map.models import (
    Assembly,
    AssemblyLayer,
    BoundaryKind,
    Fragment,
    RoutedUnit,
    Scenario,
)
from perk_dev.prose_review.catalog import AssemblyLayerView, CatalogSnapshot
from perk_dev.prose_review.source_adapter import (
    RangeFailure,
    SourceExtraction,
    SourceReadError,
    UnresolvedRange,
    read_unit_file,
    source_adapter_for,
)
from perk_dev.prose_review.source_adapter.typescript import (
    TypeScriptAdapterUnavailable,
    TypeScriptSourceAdapter,
)

type BoundaryOwner = Literal["pi", "user", "runtime", "borrowed-package"]
type LayerPresence = Literal["always", "varies"]
type PresentationControl = Literal["ambient", "tools"]
type OwnedContentKind = Literal["rendered-template", "raw-source", "source-fragments"]
type AssemblyLayerFailureReason = (
    RangeFailure
    | Literal[
        "source-unavailable",
        "template-grammar-invalid",
        "template-include-unsupported",
        "template-variable-unknown",
        "template-render-failed",
        "unsupported-family",
        "adapter-unavailable",
    ]
)
type AssemblyRenderErrorReason = Literal[
    "unknown-assembly",
    "unknown-scenario",
    "scenario-assembly-mismatch",
    "duplicate-workspace-path",
    "unknown-workspace-path",
]

# The exact descriptive marker for authored `optional: true` layers: presence varies by
# consuming session shape or runtime state — never a render-time filter.
PRESENCE_VARIES_LABEL = "Presence varies by session shape or runtime."

# The closed BoundaryKind → semantic owner id mapping (placeholders carry owner AND kind).
BOUNDARY_OWNERS: dict[BoundaryKind, BoundaryOwner] = {
    "pi-system": "pi",
    "user-content": "user",
    "runtime-state": "runtime",
    "borrowed-prompt": "borrowed-package",
}

# Exact safe per-layer failure details. No resolved paths, OS errors, adapter diagnostics,
# helper protocol details, or raw exception text ever enter an Assembly result.
FAILURE_DETAILS: dict[AssemblyLayerFailureReason, str] = {
    "source-unavailable": "The canonical source could not be read safely.",
    "template-grammar-invalid": (
        "The prompt template uses syntax outside the supported preview grammar."
    ),
    "template-include-unsupported": "Assembly preview does not support prompt includes.",
    "template-variable-unknown": (
        "The prompt template references a name outside the scenario's variables."
    ),
    "template-render-failed": "The prompt template could not be rendered for this scenario.",
    "unsupported-family": "The source family has no assembly extraction adapter.",
    "adapter-unavailable": "The source adapter could not run safely.",
    "unsupported-selector": "A catalog fragment uses a selector unsupported by its source adapter.",
    "unsupported-source-shape": (
        "A catalog fragment resolves to a source shape that cannot be extracted safely."
    ),
    "selector-not-found": "A catalog fragment no longer resolves in the current source.",
    "selector-ambiguous": "A catalog fragment resolves more than once in the current source.",
    "invalid-source": "The current source is not syntactically valid for its adapter.",
}


@dataclass(frozen=True, slots=True)
class WorkspaceBuffer:
    path: str
    text: str


@dataclass(frozen=True, slots=True)
class PresentationOverrides:
    include_ambient: bool | None
    include_tools: bool | None


@dataclass(frozen=True, slots=True)
class ResolvedPresentation:
    include_ambient: bool
    include_tools: bool


@dataclass(frozen=True, slots=True)
class LayerPresentation:
    position: int
    label: str | None
    presence: LayerPresence
    presence_label: str | None
    visibility_control: PresentationControl | None


@dataclass(frozen=True, slots=True)
class RenderedContentPart:
    fragment: Fragment | None
    text: str


@dataclass(frozen=True, slots=True)
class AssemblyLayerProblem:
    fragment: Fragment | None
    reason: AssemblyLayerFailureReason
    detail: str


@dataclass(frozen=True, slots=True)
class RenderedOwnedLayer:
    presentation: LayerPresentation
    unit: RoutedUnit
    content_kind: OwnedContentKind
    parts: tuple[RenderedContentPart, ...]


@dataclass(frozen=True, slots=True)
class RenderedBoundaryLayer:
    presentation: LayerPresentation
    boundary: BoundaryKind
    owner: BoundaryOwner


@dataclass(frozen=True, slots=True)
class FailedAssemblyLayer:
    presentation: LayerPresentation
    unit: RoutedUnit
    problems: tuple[AssemblyLayerProblem, ...]


@dataclass(frozen=True, slots=True)
class RenderedAssembly:
    assembly: Assembly
    scenario: Scenario
    presentation: ResolvedPresentation
    layers: tuple[RenderedOwnedLayer | RenderedBoundaryLayer | FailedAssemblyLayer, ...]


class AssemblyRenderError(Exception):
    """The whole render request is invalid, with a closed reason (never a layer result)."""

    def __init__(self, reason: AssemblyRenderErrorReason) -> None:
        super().__init__(reason)
        self.reason: AssemblyRenderErrorReason = reason


def _problem(
    fragment: Fragment | None,
    reason: AssemblyLayerFailureReason,
) -> AssemblyLayerProblem:
    return AssemblyLayerProblem(fragment=fragment, reason=reason, detail=FAILURE_DETAILS[reason])


def _layer_presentation(view: AssemblyLayerView) -> LayerPresentation:
    layer: AssemblyLayer = view.layer
    control: PresentationControl | None = None
    if view.unit is not None:
        if view.unit.role == "ambient-discovery":
            control = "ambient"
        elif view.unit.role == "tool-contract":
            control = "tools"
    return LayerPresentation(
        position=view.position,
        label=layer.label,
        presence="varies" if layer.optional else "always",
        presence_label=PRESENCE_VARIES_LABEL if layer.optional else None,
        visibility_control=control,
    )


def _is_raw_markdown(unit: RoutedUnit) -> bool:
    """Markdown outside ``prompts/`` (skills, agent definitions, managed ``.md`` prose)."""
    return (
        unit.candidate.kind in ("markdown", "managed-prose")
        and Path(unit.candidate.path).suffix.lower() == ".md"
    )


class AssemblyRenderer:
    """App-lifetime assembly preview composer over an explicit per-call snapshot.

    Captures only the resolved source root and the already-created app-scoped TypeScript
    SourceAdapter (its helper slot stays a real cross-operation resource bound; the
    concrete type also owns the ``TypeScriptAdapterUnavailable`` contract this renderer
    maps to the typed ``adapter-unavailable`` failure — a broader adapter here could
    misreport a wiring error as a content failure). The renderer accepts no session-shape
    id, request-selected read path, selector, fragment list, adapter, helper root,
    scenario variable override, or executable callback.
    """

    def __init__(self, repo_root: Path, typescript_adapter: TypeScriptSourceAdapter) -> None:
        self._repo_root = repo_root.resolve()
        self._typescript_adapter = typescript_adapter

    def render(
        self,
        snapshot: CatalogSnapshot,
        *,
        assembly_id: str,
        scenario_id: str,
        presentation: PresentationOverrides,
        workspace_buffers: tuple[WorkspaceBuffer, ...],
    ) -> RenderedAssembly:
        """Render every authored layer of one assembly under one selected scenario."""
        view = snapshot.get_assembly(assembly_id)
        if view is None:
            raise AssemblyRenderError("unknown-assembly")
        scenario = snapshot.get_scenario(scenario_id)
        if scenario is None:
            raise AssemblyRenderError("unknown-scenario")
        if scenario.assembly != assembly_id:
            raise AssemblyRenderError("scenario-assembly-mismatch")
        # The request is the browser's complete loaded workspace, so catalog-known paths
        # unrelated to this assembly are valid; duplicates and unknown paths are not.
        buffers: dict[str, str] = {}
        for buffer in workspace_buffers:
            if buffer.path in buffers:
                raise AssemblyRenderError("duplicate-workspace-path")
            if not snapshot.units_for_path(buffer.path):
                raise AssemblyRenderError("unknown-workspace-path")
            buffers[buffer.path] = buffer.text
        resolved = ResolvedPresentation(
            include_ambient=(
                scenario.include_ambient
                if presentation.include_ambient is None
                else presentation.include_ambient
            ),
            include_tools=(
                scenario.include_tools
                if presentation.include_tools is None
                else presentation.include_tools
            ),
        )
        # Request-local workspace-first text resolution: a buffered path (even an empty
        # string) is never reread; canonical reads and read failures are shared per path.
        text_cache: dict[str, str | None] = dict(buffers)
        layers: list[RenderedOwnedLayer | RenderedBoundaryLayer | FailedAssemblyLayer] = []
        for layer_view in view.layers:
            layers.append(self._render_layer(layer_view, scenario, text_cache))
        return RenderedAssembly(
            assembly=view.assembly,
            scenario=scenario,
            presentation=resolved,
            layers=tuple(layers),
        )

    def _render_layer(
        self,
        view: AssemblyLayerView,
        scenario: Scenario,
        text_cache: dict[str, str | None],
    ) -> RenderedOwnedLayer | RenderedBoundaryLayer | FailedAssemblyLayer:
        presentation = _layer_presentation(view)
        boundary = view.layer.boundary
        if boundary is not None:
            return RenderedBoundaryLayer(
                presentation=presentation,
                boundary=boundary,
                owner=BOUNDARY_OWNERS[boundary],
            )
        unit = view.unit
        if unit is None:  # pragma: no cover - a clean catalog authors one unit or boundary
            raise AssertionError("assembly layer carries neither unit nor boundary")
        text = self._layer_text(unit, text_cache)
        if text is None:
            return FailedAssemblyLayer(
                presentation=presentation,
                unit=unit,
                problems=(_problem(None, "source-unavailable"),),
            )
        template_name = prompt_template_name(unit.candidate)
        if template_name is not None:
            return self._render_prompt_layer(presentation, unit, scenario, text)
        if _is_raw_markdown(unit):
            return RenderedOwnedLayer(
                presentation=presentation,
                unit=unit,
                content_kind="raw-source",
                parts=(RenderedContentPart(fragment=None, text=text),),
            )
        return self._render_code_layer(presentation, unit, text)

    def _layer_text(self, unit: RoutedUnit, text_cache: dict[str, str | None]) -> str | None:
        path = unit.candidate.path
        if path in text_cache:
            return text_cache[path]
        try:
            text: str | None = read_unit_file(self._repo_root, unit).text
        except SourceReadError:
            text = None
        text_cache[path] = text
        return text

    def _render_prompt_layer(
        self,
        presentation: LayerPresentation,
        unit: RoutedUnit,
        scenario: Scenario,
        text: str,
    ) -> RenderedOwnedLayer | FailedAssemblyLayer:
        def failed(reason: AssemblyLayerFailureReason) -> FailedAssemblyLayer:
            return FailedAssemblyLayer(
                presentation=presentation,
                unit=unit,
                problems=(_problem(None, reason),),
            )

        scan = scan_template(text)
        if scan.violations:
            return failed("template-grammar-invalid")
        if scan.has_include:
            # Assembly preview supports no include (canonical or workspace): no editable
            # request may trigger the packaged prompts_dir() loader, and no render may
            # combine this renderer's source root with packaged/checkout prompt bytes.
            return failed("template-include-unsupported")
        variables = dict(scenario.variables)
        if scan.identifiers - variables.keys():
            # Mapping-only identifier semantics (the TS mini-jinja twin's namespace):
            # jinja's default Environment globals and the true/false/none literals are
            # unreachable because none appears in an authored scenario mapping.
            return failed("template-variable-unknown")
        try:
            rendered = render_text(text, variables)
        except TemplateError:
            # Reachable through gate-passing structural errors (if/endif imbalance).
            # A TypeError (non-string variable) stays loud: scenario variables are
            # catalog-validated string pairs, so that is a broken internal invariant.
            return failed("template-render-failed")
        return RenderedOwnedLayer(
            presentation=presentation,
            unit=unit,
            content_kind="rendered-template",
            parts=(RenderedContentPart(fragment=None, text=rendered),),
        )

    def _render_code_layer(
        self,
        presentation: LayerPresentation,
        unit: RoutedUnit,
        text: str,
    ) -> RenderedOwnedLayer | FailedAssemblyLayer:
        adapter = source_adapter_for(unit, typescript_adapter=self._typescript_adapter)
        if adapter is None:
            return FailedAssemblyLayer(
                presentation=presentation,
                unit=unit,
                problems=(_problem(None, "unsupported-family"),),
            )
        fragments = unit.candidate.fragments
        selectors = tuple(fragment.selector for fragment in fragments)
        try:
            extractions = adapter.extract_many(text, selectors)
        except TypeScriptAdapterUnavailable:
            return FailedAssemblyLayer(
                presentation=presentation,
                unit=unit,
                problems=(_problem(None, "adapter-unavailable"),),
            )
        problems = _code_problems(fragments, extractions)
        if problems:
            # Code layers are atomic result variants: any unresolved fragment fails the
            # whole authored layer with ordered problems and no partial content.
            return FailedAssemblyLayer(
                presentation=presentation,
                unit=unit,
                problems=problems,
            )
        return RenderedOwnedLayer(
            presentation=presentation,
            unit=unit,
            content_kind="source-fragments",
            parts=tuple(
                RenderedContentPart(fragment=fragment, text=extraction.focus)
                for fragment, extraction in zip(fragments, extractions, strict=True)
            ),
        )


def _code_problems(
    fragments: tuple[Fragment, ...],
    extractions: tuple[SourceExtraction, ...],
) -> tuple[AssemblyLayerProblem, ...]:
    """Ordered per-fragment problems, collapsing document-level invalid-source to one."""
    problems: list[AssemblyLayerProblem] = []
    saw_invalid_source = False
    for fragment, extraction in zip(fragments, extractions, strict=True):
        resolution = extraction.resolution
        if not isinstance(resolution, UnresolvedRange):
            continue
        if resolution.reason == "invalid-source":
            if not saw_invalid_source:
                problems.append(_problem(None, "invalid-source"))
                saw_invalid_source = True
            continue
        problems.append(_problem(fragment, resolution.reason))
    return tuple(problems)

File: tools/prose-review/src/assembly.ts
// The typed fetch boundary for /api/assembly/options and /api/assembly/render: local
// mirrors of the wire shapes plus reject-unknown structural parsers (the
// tree.ts/comparison.ts posture — every known field required and correctly typed; any
// defect rejects the whole payload with null). The closed vocabularies below are
// endpoint-only, so they live here rather than in wire.ts. One deliberate opening:
// a problem's `reason` is validated as a non-empty string, not a closed set — the
// client renders the server's fixed `detail` copy and must not duplicate the
// server-internal reason vocabulary.

import { type FragmentRef, parseFragmentRef } from "./selection.ts";
import { parseUnitRef, type UnitRef } from "./tree.ts";
import { type BoundaryKind, isBoundaryKind } from "./wire.ts";

export const LAYER_PRESENCES = ["always", "varies"] as const;

export type LayerPresence = (typeof LAYER_PRESENCES)[number];

export const PRESENTATION_CONTROLS = ["ambient", "tools"] as const;

export type PresentationControl = (typeof PRESENTATION_CONTROLS)[number];

export const BOUNDARY_OWNERS = ["pi", "user", "runtime", "borrowed-package"] as const;

export type BoundaryOwner = (typeof BOUNDARY_OWNERS)[number];

export const OWNED_CONTENT_KINDS = ["rendered-template", "raw-source", "source-fragments"] as const;

export type OwnedContentKind = (typeof OWNED_CONTENT_KINDS)[number];

export type AssemblyScenario = {
  id: string;
  label: string;
  variables: Record<string, string>;
  include_ambient: boolean;
  include_tools: boolean;
};

// The parse boundary guarantees at least one scenario, so scenarios[0] is defined
// under noUncheckedIndexedAccess and the session controller stays total.
export type AssemblyOptions = {
  assembly: string;
  scenarios: [AssemblyScenario, ...AssemblyScenario[]];
};

export type AssemblyPresentation = {
  include_ambient: boolean;
  include_tools: boolean;
};

export type AssemblyLayerPresentation = {
  position: number;
  label: string | null;
  presence: LayerPresence;
  presence_label: string | null;
  visibility_control: PresentationControl | null;
};

export type AssemblyContentPart = {
  fragment: FragmentRef | null;
  text: string;
};

export type AssemblyLayerProblem = {
  fragment: FragmentRef | null;
  reason: string;
  detail: string;
};

export type AssemblyOwnedLayer = {
  type: "owned";
  presentation: AssemblyLayerPresentation;
  unit: UnitRef;
  content_kind: OwnedContentKind;
  parts: AssemblyContentPart[];
};

export type AssemblyBoundaryLayer = {
  type: "boundary";
  presentation: AssemblyLayerPresentation;
  boundary: BoundaryKind;
  owner: BoundaryOwner;
};

export type AssemblyFailureLayer = {
  type: "failure";
  presentation: AssemblyLayerPresentation;
  unit: UnitRef;
  problems: AssemblyLayerProblem[];
};

export type AssemblyRenderedLayer = AssemblyOwnedLayer | AssemblyBoundaryLayer | AssemblyFailureLayer;

export type AssemblyRender = {
  assembly: string;
  scenario: AssemblyScenario;
  presentation: AssemblyPresentation;
  layers: AssemblyRenderedLayer[];
};

// The exact render request identity (the nullable overrides ride verbatim so the
// server echo matches the request-time resolution).
export type AssemblyRenderRequest = {
  assembly: string;
  scenario: string;
  presentation: {
    include_ambient: boolean | null;
    include_tools: boolean | null;
  };
};

export type AssemblyOverrides = {
  ambient: boolean | null;
  tools: boolean | null;
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function nonEmpty(value: unknown): value is string {
  return typeof value === "string" && value.trim().length > 0;
}

function parseArray<T>(value: unknown, parseEntry: (entry: unknown) => T | null): T[] | null {
  if (!Array.isArray(value)) {
    return null;
  }
  const entries: T[] = [];
  for (const entry of value) {
    const parsed = parseEntry(entry);
    if (parsed === null) {
      return null;
    }
    entries.push(parsed);
  }
  return entries;
}

function isLayerPresence(value: unknown): value is LayerPresence {
  return typeof value === "string" && (LAYER_PRESENCES as readonly string[]).includes(value);
}

function isPresentationControl(value: unknown): value is PresentationControl {
  return typeof value === "string" && (PRESENTATION_CONTROLS as readonly string[]).includes(value);
}

function isBoundaryOwner(value: unknown): value is BoundaryOwner {
  return typeof value === "string" && (BOUNDARY_OWNERS as readonly string[]).includes(value);
}

function isOwnedContentKind(value: unknown): value is OwnedContentKind {
  return typeof value === "string" && (OWNED_CONTENT_KINDS as readonly string[]).includes(value);
}

// Entries stay in received key order — the server sorts, the client only displays.
function parseVariables(value: unknown): Record<string, string> | null {
  if (!isRecord(value) || Array.isArray(value)) {
    return null;
  }
  const variables: Record<string, string> = {};
  for (const [key, entry] of Object.entries(value)) {
    if (typeof entry !== "string") {
      return null;
    }
    variables[key] = entry;
  }
  return variables;
}

function parseScenario(value: unknown): AssemblyScenario | null {
  if (
    !isRecord(value) ||
    !nonEmpty(value.id) ||
    typeof value.label !== "string" ||
    typeof value.include_ambient !== "boolean" ||
    typeof value.include_tools !== "boolean"
  ) {
    return null;
  }
  const variables = parseVariables(value.variables);
  if (variables === null) {
    return null;
  }
  return {
    id: value.id,
    label: value.label,
    variables,
    include_ambient: value.include_ambient,
    include_tools: value.include_tools,
  };
}

/**
 * Validate one complete assembly-options response; null rejects any malformed member.
 * An empty `scenarios` array is rejected too: the backend's scenario-data completeness
 * makes it an ill-shaped payload, not a real state.
 */
export function parseAssemblyOptions(value: unknown): AssemblyOptions | null {
  if (!isRecord(value) || !nonEmpty(value.assembly)) {
    return null;
  }
  const scenarios = parseArray(value.scenarios, parseScenario);
  if (scenarios === null) {
    return null;
  }
  const [first, ...rest] = scenarios;
  if (first === undefined) {
    return null;
  }
  return { assembly: value.assembly, scenarios: [first, ...rest] };
}

function parsePresentation(value: unknown): AssemblyPresentation | null {
  if (
    !isRecord(value) ||
    typeof value.include_ambient !== "boolean" ||
    typeof value.include_tools !== "boolean"
  ) {
    return null;
  }
  return { include_ambient: value.include_ambient, include_tools: value.include_tools };
}

function parseLayerPresentation(value: unknown): AssemblyLayerPresentation | null {
  if (!isRecord(value)) {
    return null;
  }
  const { position, label, presence_label: presenceLabel } = value;
  if (typeof position !== "number" || !Number.isInteger(position) || position < 1) {
    return null;
  }
  if (label !== null && typeof label !== "string") {
    return null;
  }
  if (!isLayerPresence(value.presence)) {
    return null;
  }
  if (presenceLabel !== null && typeof presenceLabel !== "string") {
    return null;
  }
  const control = value.visibility_control;
  if (control !== null && !isPresentationControl(control)) {
    return null;
  }
  return {
    position,
    label,
    presence: value.presence,
    presence_label: presenceLabel,
    visibility_control: control,
  };
}

function parseNullableFragment(value: unknown): { fragment: FragmentRef | null } | null {
  if (value === null) {
    return { fragment: null };
  }
  const fragment = parseFragmentRef(value);
  if (fragment === null) {
    return null;
  }
  return { fragment };
}

function parsePart(value: unknown): AssemblyContentPart | null {
  if (!isRecord(value) || typeof value.text !== "string") {
    return null;
  }
  const fragment = parseNullableFragment(value.fragment);
  if (fragment === null) {
    return null;
  }
  return { fragment: fragment.fragment, text: value.text };
}

function parseProblem(value: unknown): AssemblyLayerProblem | null {
  if (!isRecord(value) || !nonEmpty(value.reason) || typeof value.detail !== "string") {
    return null;
  }
  const fragment = parseNullableFragment(value.fragment);
  if (fragment === null) {
    return null;
  }
  return { fragment: fragment.fragment, reason: value.reason, detail: value.detail };
}

function parseLayer(value: unknown): AssemblyRenderedLayer | null {
  if (!isRecord(value)) {
    return null;
  }
  const presentation = parseLayerPresentation(value.presentation);
  if (presentation === null) {
    return null;
  }
  if (value.type === "owned") {
    const unit = parseUnitRef(value.unit);
    const parts = parseArray(value.parts, parsePart);
    if (unit === null || parts === null || !isOwnedContentKind(value.content_kind)) {
      return null;
    }
    return { type: "owned", presentation, unit, content_kind: value.content_kind, parts };
  }
  if (value.type === "boundary") {
    if (!isBoundaryKind(value.boundary) || !isBoundaryOwner(value.owner)) {
      return null;
    }
    return { type: "boundary", presentation, boundary: value.boundary, owner: value.owner };
  }
  if (value.type === "failure") {
    const unit = parseUnitRef(value.unit);
    const problems = parseArray(value.problems, parseProblem);
    if (unit === null || problems === null) {
      return null;
    }
    return { type: "failure", presentation, unit, problems };
  }
  return null;
}

/** Validate one complete assembly-render response; null rejects any malformed member. */
export function parseAssemblyRender(value: unknown): AssemblyRender | null {
  if (!isRecord(value) || !nonEmpty(value.assembly)) {
    return null;
  }
  const scenario = parseScenario(value.scenario);
  const presentation = parsePresentation(value.presentation);
  const layers = parseArray(value.layers, parseLayer);
  if (scenario === null || presentation === null || layers === null) {
    return null;
  }
  return { assembly: value.assembly, scenario, presentation, layers };
}

/** The comparisonOptionsMatchRequest posture: the echo must name the requested subject. */
export function assemblyRenderMatchesRequest(
  render: AssemblyRender,
  request: AssemblyRenderRequest,
): boolean {
  return render.assembly === request.assembly && render.scenario.id === request.scenario;
}

/** Resolve the two visibility booleans locally: `override ?? scenario default`. */
export function resolvedPresentation(
  scenario: AssemblyScenario,
  overrides: AssemblyOverrides,
): AssemblyPresentation {
  return {
    include_ambient: overrides.ambient ?? scenario.include_ambient,
    include_tools: overrides.tools ?? scenario.include_tools,
  };
}

// Visibility is derived client-side and never re-POSTs: a layer is hidden exactly when
// its visibility_control names a control whose resolved value is false. Presence never
// affects visibility — it is display metadata only.
export function visibleLayers(
  layers: AssemblyRenderedLayer[],
  resolved: AssemblyPresentation,
): AssemblyRenderedLayer[] {
  return layers.filter((layer) => {
    const control = layer.presentation.visibility_control;
    if (control === "ambient") {
      return resolved.include_ambient;
    }
    if (control === "tools") {
      return resolved.include_tools;
    }
    return true;
  });
}

function layerText(layer: AssemblyRenderedLayer): string {
  if (layer.type === "owned") {
    return layer.parts.map((part) => part.text).join("");
  }
  if (layer.type === "boundary") {
    const label = layer.presentation.label ?? layer.boundary;
    return `[[ boundary: ${label} · owner: ${layer.owner} ]]`;
  }
  return `[[ layer failed: ${layer.unit.id} ]]`;
}

/** Join currently-visible layers in delivery order with one blank line between layers. */
export function concatenatedText(layers: AssemblyRenderedLayer[]): string {
  return layers.map(layerText).join("\n\n");
}
